sycl/doc/extensions/proposed/sycl_ext_oneapi_annotated_ptr.asciidoc

```diff
@@ -0,0 +1,510 @@
+----
+|
+Allows assignment from an pointer to type `T`.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+annotated_ptr& operator=(annotated_ptr const&) noexcept;
+----
+|
+Allows assignment from an `annotated_ptr` with the same parameterization.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+annotated_ptr& operator++() noexcept;
+----
+|
+Prefix increment operator.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+annotated_ptr operator++() noexcept;
+----
+|
+Postfix increment operator.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+annotated_ptr& operator--() noexcept;
+----
+|
+Prefix decrement operator.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+annotated_ptr operator--() noexcept;
+----
+|
+Postfix decrement operator.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+template<typename propertyT>
+static constexpr bool has_property();
+----
+| Returns true if the `PropertyListT` contains the property specified by `propertyT`. Returns false if it does not.
+Available only if `sycl::is_property_of_v<propertyT, sycl::ext::oneapi::annotated_ptr>` is true.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+template<typename propertyT>
+static constexpr auto get_property();
+----
+| Returns an object of the class used to represent the value of property `propertyT`.
+Must produce a compiler diagnostic if `PropertyListT` does not contain a `propertyT` property.
+Available only if `sycl::is_property_of_v<propertyT, sycl::ext::oneapi::annotated_ptr>` is true.
+|===
+
+=== Add new reference wrapper class `annotated_ref` to enable `annotated_ptr`
+
+The purpose of the `annotated_ref` class template is to provide reference wrapper semantics. It enables the implementation to preserve the properties on loads from and stores to the pointers.
+
+```c++
+namespace sycl::ext::oneapi {
+template <typename T, typename PropertyListT = property_list_t<>>
+class annotated_ref {
+  public:
+    annotated_ref(T *);
+    operator T() noexcept;
+    operator const T() const noexcept;
+    void operator=(const T &);
+  private:
+    T *Ptr;
+  };
+} // namespace sycl::ext::oneapi
+```
+
+
+Member Functions are described in the table below
+[frame="topbot",options="header"]
+|===
+|Functions |Description
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+annotated_ref(T * InputPtr);
+----
+|
+Constructs an `annotated_ref` object. Does not allocate new storage. The underlying pointer is initialized to `InputPtr`.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+operator T() noexcept;
+operator const T() const noexcept;
+----
+|
+Implicit conversion to underlying type.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+void operator=(const T &);
+----
+|
+Enables assignment to the underlying pointer.
+
+|===
+
+=== Properties for `annotated_ptr` variables
+
+See the extension `sycl_ext_oneapi_annotated_ptr_properties` for the full list of supported properties.
+
+== Issues
+
+1) [RESOLVED] Should we allow implicit conversion to base class by default?
+*Ans: No.*
```

Comment:
I think it would be useful to give a bit more information here for posterity.  If we just say "No" then somebody might want to reopen the question.  Just a one line explanation like "It's too error-prone" or "Users may accidentally lose compile-time properties" would be sufficient.

It's also unclear to me if this issue is related to `annotated_ptr`, `annotated_ref` or both.

Reply:
Added more info. Added clarification in section title.